### ISSUE TITLE
How to set up a CustomUserRepository?

### ISSUE THREAD

**Issue by Vincent026** [brockallen/BrockAllen.MembershipReboot#701]
Hello,

we want to make a CustomUserRepository based on our own database repository (no EF, Mongo or Raven). I derive this repo from QueryableUserAccountRepository, register the user & group repo's (as singletons) plus their services with the IdentityManagerServiceFactory and implement the methods. I have some trouble to make it work. I manage to show all users but when I try to edit a single user I get _Error: An error has occurred._ 

Some useful hints would be great.

thanks,
Vincent

**Comment by brockallen** [comment]
Maybe this helps: https://github.com/brockallen/BrockAllen.MembershipReboot/wiki/Implementing-A-Custom-Storage-Mechanism

**Comment by Vincent026** [comment]
Thanks for the quick response Brock! I will have a close look at this article again.

**Comment by Vincent026** [comment]
Our current implementation, based on your examples, uses some InMemory clients, users and scopes, registered in the IdentityServerServiceFactory, which is mapped to the path "/identity". I see that there is also a mapped path named "/core". With which mapping do I need to use to the configure the IdentityServerServiceFactory in order to register the custom user service?

kind regards

**Comment by Vincent026** [comment]
My implementation of MembershipReboot seems to work fine. I choose to implement the interfaces instead of deriving from QueryableUserAccountRepository. I have better insight in the way CustomUser instances are created now. 
But I still have some trouble implementing IdentityServer. I follow the setup from IdentityServer3.Samples/source/MembershipReboot/. I get the response "invalid_grant" when calling RequestResourceOwnerPasswordAsync.  Any ideas?

**Comment by brockallen** [comment]
Whichever you want. All of those are just samples.

**Comment by Vincent026** [comment]
Thanks! My first thought was that these where compulsory predefined end points. I am getting closer to what I want.

**Comment by Vincent026** [comment]
I am making progress with implementing CustomUserAccountService and CustomUserRepository. It seems that a hashed password is expected to be stored. Our table has a readable password field and that is ok with us. I rather don't modify the schema of the user table, What I am trying to do now is intercepting authentication by overriding the Authenticate method(s) in the CustomUserAccountService. I have two questions on that:
1. is this a good strategy for overriding authentication (to validate the password with the readable field instead of using the hashedpassword)?
2. do I need the store a reference to the repo in CustomUserAccountService, or is there an existing field/property in the baseclass (UserAccountService) that returns the repo? I couldn't find one.

**Comment by brockallen** [comment]
If you have an existing DB for users, then I don't see why you want to use MR. Just implement the IUserService in IdentityServer.

**Comment by Vincent026** [comment]
You are right that we already have a way of filling our DB and there is no need for us to have MR to do the CRUD actions. However, this may change in the future. Besides that, we have several user sources and I can use MR to show the combined set. I don't implement the Create, Update and Delete methods in the repo.
